tests/test_commands_cache.py

```diff
@@ -26,6 +29,32 @@ def test_predict_threadable_unknown_command(xonsh_builtins):
     assert isinstance(result, bool)
 
 
+def test_commands_cached_between_runs(xonsh_builtins, tmp_path, monkeypatch):
```

Comment:
This test checks that a pickle file was created, but it doesn't check that it will be used in a new `CommandsCache()` instance

Reply:
> Hey @jnoortheen - I've looked at this some more and it seems that **all code in xonsh depends on `all_commands` updating the cache instantly**, so updating it in a background thread seems to break that.
> Also, calling it multiple times will cause multiple threads to be created unnecessarily.
> Another issue is that even if the pickled value is updated correctly, you still create a thread.
> 
> I have an alternative suggestion:
> In the same pickle file, save `self._path_checksum`, `self._alias_checksum` and `self._path_mtime`.
> In `all_commands`, load these attributes from the pickle file before checking the new values in `_update_if_changed`.
> 
> This way the rest of `all_commands` stays the same and its behavior doesn't break.

The very first time when this setting is enabled, it will return empty results. Usually from the call of xonsh-pygments plugin. This is just a quick implementation. It doesn't use locks. If there is more calls when a thread is running, it will create new ones. But for this case, it is just a quick better option. We can improve it in another PR.

atmost even if there is no change, it will create a thread at the start of the program to update the pickle file. Next time onwards the function will see that there was no change and will just return the all_commands stored in the variable.

Reply:
> This test checks that a pickle file was created, but it doesn't check that it will be used in a new CommandsCache() instance

added another test that loads result from pickle file if available and checks the results.